Replaced obsolete JEP-12 with JEP-12-01. (#138)
jmespath-unofficial/jmespath.jep: jep-012a-raw-string-literals.md

# Raw String Literals

|||
|---|---
| **JEP**    | 12a
| **Author** | Michael Downling, Maxime Labelle, Richard Gibson
| **Status** | accepted
| **Created**| 19-Nov-2022
| **Obsoletes**| [JEP-12](./jep-012-raw-string-literals.md)

## Abstract

This JEP proposes the following modifications to JMESPath in order to improve
the usability of the language and ease the implementation of parsers:

* Addition of a **raw string literal** to JMESPath that will allow expressions
to contain raw strings that are not mutated by JSON escape sequences (e.g.,

> `"\\n"`, `"\\r"`, `"\\u005C"`).

Reviewer: I don't think double-quoted strings make sense in the paragraph introducing single-quoted strings.
```suggestion
* Addition of a **raw string literal** to JMESPath that will allow direct
expression of string contents that would otherwise be modified by
interpretation as JSON (e.g., `'\n'`, `'\r'`, `'\u005C'`).
```

* Deprecation of  the current literal parsing behavior that allows for unquoted
JSON strings to be parsed as JSON strings, removing an ambiguity in the
JMESPath grammar and helping to ensure consistency among implementations.

This proposal seeks to add the following syntax to JMESPath:

```
'foobar'
'foo\'bar'

> `bar` -> Parse error/warning (implementation specific)

Reviewer: ```suggestion
`bar` -> Parse error
```

```

## Motivation

Raw string literals are provided in [various programming languages](https://en.wikipedia.org/wiki/String_literal#Raw_strings) in order to prevent
language specific interpretation (i.e., JSON parsing) and remove the need for
escaping, avoiding a common problem called [leaning toothpick syndrome (LTS)](https://en.wikipedia.org/wiki/Leaning_toothpick_syndrome). Leaning toothpick
syndrome is an issue in which strings become unreadable due to excessive use of

> escape characters in order to avoid delimiter collision (e.g., `\\\\\\`).

Reviewer: ```suggestion
escape characters in order to avoid delimiter collision (e.g., `"\\\\\\"`).
```

When evaluating a JMESPath expression, it is often necessary to utilize string
literals that are not extracted from the data being evaluated, but rather
statically part of the compiled JMESPath expression. String literals are useful
in many areas, but most notably when invoking functions or building up
multi-select lists and hashes.

> The following expression produces the three-character string `"foo"` using a `` `…` `` JSON text literal:

Reviewer: ```suggestion
The following expression produces the three-character string “foo” using a `` `…` `` JSON text literal:
```

Author: Good catch.

5. If `$temp` cannot be parsed as valid JSON, then wrap the contents of
`$lexeme` in double quotes and parse the wrapped value as a JSON string,
making the following expressions equivalent: `` `foo` `` == `` `"foo"` ``, and
`` `[1, ]` `` == `` `"[1, ]"` ``.

It is reasonable to assume that the most common use case for a JSON literal in
a JMESPath expression is to provide a string value to a function argument or
to provide a literal string value to a value in a multi-select list or
multi-select hash. In order to make providing string values easier, it was
decided that JMESPath should allow the quotes around the string to be elided.

This proposal posits that allowing quotes to be elided when parsing JSON
literals should be prohibited in favor of the proper string literal syntax.

## Specification

A raw string literal is value that begins and ends with a single quote
and preserves embedded backslashes except those used to escape
backslash or single quote characters.

### Examples

Here are several examples of valid raw string literals and how they are
parsed:

* A basic raw string literal, representing the seven-character string “foo bar”:

```
'foo bar'
```

> * A raw string literal with an escaped single quote, representing the seven-character string “foo'bar”`:

Reviewer: ```suggestion
* A raw string literal with an escaped single quote, representing the seven-character string “foo'bar”:
```

```
'foo\'bar'
```

* A raw string literal with an escaped backslash character, representing the seven-character string “foo\bar”:

```
'foo\\bar'
```

> * A raw string literal that contains new lines:
> 
> ```
> 'foo
> bar
> baz!'
> ```
> 
> The above expression represents the multi-line string:
> 
> ```
> foo
> baz
> bar!
> ```

Reviewer: `````suggestion
* A raw string literal that contains new lines:

```
'foo
bar
baz!'
```

The above expression represents the multi-line string:

```
foo
bar
baz!
```
`````

> * A raw string literal that contains escape characters, representing the eight-character string “foo\nbar”:

Reviewer: ```suggestion
* A raw string literal that contains a preserved backslash character, representing the eight-character string “foo\nbar”:
```

```
'foo\nbar'
```

### ABNF

> The following ABNF grammar rules will be added, and is allowed anywhere an
> expression is allowed:
> 
> ```abnf
> raw-string = "'" *raw-string-char "'"

Reviewer: `````suggestion
The following ABNF grammar rules will be added:

```abnf
expression =/ raw-string
raw-string = "'" *raw-string-char "'"
`````

> raw-string-char = (%x00-26 /            ; ␀ - & precedes U+0027 "'" apostrophe
>                     %x28-5B /           ; ( - [ precedes U+005C "\" reverse solidus
>                     %x5D-10FFFF) /      ; ] - ... "]" and all following code points
>                     preserved-escape /
>                     raw-string-escape
> preserved-escape = escape (
>                     %x00-26 /           ;  ␀ - & precedes U+0027 "'" apostrophe
>                     %x28-5B /           ; ( -[ precedes U+005C "\" reverse solidus
>                     %x5D-10FFFF)        ; ] - ... "]" and all following code points

Reviewer: ```suggestion
raw-string-char = (%x00-26 /            ; ␀ through '&' (precedes U+0027 APOSTROPHE "'")
                    %x28-5B /           ; '(' through '[' (precedes U+005C REVERSE SOLIDUS '\')
                    %x5D-10FFFF) /      ; ']' and all following code points
                    preserved-escape /
                    raw-string-escape
preserved-escape = escape (
                    %x00-26 /           ;  ␀ through '&' (precedes U+0027 APOSTROPHE "'")
                    %x28-5B /           ; '(' through '[' (precedes U+005C REVERSE SOLIDUS '\')
                    %x5D-10FFFF)        ; ']' and all following code points
```

raw-string-escape = escape (

>                     "'" /               ; \ apostrophe U+0027
>                     escape)             ; \ reverse solidus U+005C

Reviewer: ```suggestion
                    "'" /               ; U+0027 APOSTROPHE "'"
                    escape)             ; U+005C REVERSE SOLIDUS '\'
```

```

> This rule allows any character inside of a raw string,

Reviewer: ```suggestion
These rules allow any character inside of a raw string,
```

including control characters and escaped single quotes or backslashes.

In addition to adding a `raw-string` rule, the `literal` rule in the ABNF
will be simplified to become:

```
literal = "`" json-text "`"
```

## Impact

The impact to existing users of JMESPath is that the use of a JSON literal
in which the quotes are elided MUST be quoted or converted to use the
raw-string rule of the grammar.

> In order to support this type of variance in JMESPath implementations, all of
> the JSON literal compliance test cases that involve elided quotes MUST be
> removed, and test cases regarding failing on invalid unquoted JSON values MUST
> not be allowed in the compliance test unless placed in a JEP 12 specific
> test suite, allowing implementations that support elided quotes in JSON
> literals to filter out the JEP 12 specific test cases.

Reviewer: ```suggestion
To accommodate legacy JMESPath implementations, all of
the JSON literal compliance test cases that involve elided quotes MUST be
removed, and test cases regarding failing on invalid unquoted JSON values
MUST NOT be allowed in the compliance test unless placed in a JEP 12 specific
test suite, allowing such implementations to filter them out.
```
